src/Microsoft.AspNet.Cors/CorsMiddleware.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Cors.Core;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.WebUtilities;
+
+namespace Microsoft.AspNet.Cors
+{
+    /// <summary>
+    /// An ASP.NET middleware for handling CORS.
+    /// </summary>
+    public class CorsMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly CorsOptions _options;
+
+        /// <summary>
+        /// Instantiates a new <see cref="CorsMiddleware"/>.
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="options"></param>
+        public CorsMiddleware(
+            [NotNull]RequestDelegate next,
+            [NotNull]CorsOptions options)
+        {
+            _next = next;
+            _options = options;
+
+            _options.PolicyProvider = _options.PolicyProvider ?? new CorsPolicyProvider();
+            _options.CorsEngine = _options.CorsEngine ?? new CorsEngine();
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            var corsRequestContext = new CorsRequestContext(context);
+
+            if (corsRequestContext.IsCorsRequest)
+            {
+                var corsPolicy = await _options.PolicyProvider.GetCorsPolicyAsync(corsRequestContext);
+
+                if (corsPolicy != null)
+                {
+                    var corsResult = _options.CorsEngine.EvaluatePolicy(corsRequestContext, corsPolicy);
+
+                    if (corsRequestContext.IsPreflight)
+                    {
+                        // Always respond to preflight requests.
+                        if (corsResult.IsValid)
+                        {
+                            context.Response.StatusCode = StatusCodes.Status200OK;
+                            WriteCorsHeaders(context, corsResult);
+                        }
+                        else
+                        {
+                            context.Response.StatusCode = StatusCodes.Status400BadRequest;
```

Comment:
The response may not necessarily always be 400. It may be the way in which we process incorrect preflight request but after read the spec for another time I think it's debatable.
http://www.w3.org/TR/cors/#resource-preflight-requests

The words used in the spec is:
> The request is outside the scope of this specification.

I tend to interrupt it as: __treat it as a normal request come in the pipeline and see how user logic deal with it__. So even we don't process it again, the return status code should be 405 method not allowed as user tends not to define how to process OPTION.

This is just my suggestion. Practically, returns 400 may not cause any trouble, because for any status out of 2XX range user agent will simply stop the processing.